### PR TITLE
br: add flag to control whether restore priv table on default

### LINKED ISSUE #36651: br: add flag to control whether restore priv table on default
## Enhancement
add flag `--with-sys-table` to control whether restore priv table on default for full/point restore.

and when restore priv table we'll skip clearing and restoring user `cloud_admin` since it's a special user on cloud.

for the feature `restore system priv tables`, ref https://github.com/pingcap/tidb/issues/35395

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: close #36651

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by IANTHEREAL** [comment]
what if user uses `br restore full --f='mysql.*' --with-sys-table=false`

**Comment by IANTHEREAL** [comment]
suggestion for incompatible  case output

```
#######################################################################
# the target cluster is not compatible with the backup data,
# br cannot restore system tables.
# you can use the following command to skip restoring system tables:
#    br restore full --with-sys-table=false
#######################################################################
```

File: br/pkg/restore/systable_restore.go [hunk 1]
```diff
@@ -44,15 +44,15 @@ var unRecoverableTable = map[string]struct{}{
...
-var sysPrivilegeTableSet = map[string]bool{
-	"user":          true, // since v1.0.0
-	"db":            true, // since v1.0.0
-	"tables_priv":   true, // since v1.0.0
-	"columns_priv":  true, // since v1.0.0
-	"default_roles": true, // since v3.0.0
-	"role_edges":    true, // since v3.0.0
-	"global_priv":   true, // since v3.0.8
-	"global_grants": true, // since v5.0.3
+var sysPrivilegeTableMap = map[string]string{
```
Comment (WangLe1321) [inline br/pkg/restore/systable_restore.go:47]: I think you can add a comment to describe the meaning of the map value.

**Comment by D3Hunter** [comment]
> what if user uses `br restore full --f='mysql.*' --with-sys-table=false`?

will run previous logic to restore tables in `mysql` except tables which are forbidden to restore, i.e. do `replace into xxx select * from xxx`

**Comment by D3Hunter** [comment]
> suggestion for incompatible case output
> 
> ```
> #######################################################################
> # the target cluster is not compatible with the backup data,
> # br cannot restore system tables.
> # you can use the following command to skip restoring system tables:
> #    br restore full --with-sys-table=false
> #######################################################################
> ```

change to:
```
#######################################################################
# the target cluster is not compatible with the backup data,
# br cannot restore system tables.
# you can remove 'with-sys-table' flag to skip restoring system tables
#######################################################################
```

**Comment by D3Hunter** [comment]
ci success https://ci.pingcap.net/blue/organizations/jenkins/br_ghpr_unit_and_integration_test/detail/br_ghpr_unit_and_integration_test/12630/pipeline

File: br/pkg/restore/systable_restore.go (hunk 1 above)
Comment (D3Hunter) [reply br/pkg/restore/systable_restore.go:51]: [fix comments](https://github.com/pingcap/tidb/pull/36652/commits/90dc2b9a4ad8affaaca4258b65ec1a1493138401)

**Comment by D3Hunter** [comment]
@3pointer @WangLe1321 ptal

**Comment by IANTHEREAL** [comment]
> > what if user uses `br restore full --f='mysql.*' --with-sys-table=false`?
> 
> will run previous logic to restore tables in `mysql` except tables which are forbidden to restore, i.e. do `replace into xxx select * from xxx`

I understand that `with-sys-table = false` has a higher priority, so that the user and permission tables are not restored

File: br/pkg/task/common.go [hunk 2]
```diff
@@ -505,6 +516,12 @@ func (cfg *Config) ParseFromFlags(flags *pflag.FlagSet) error {
 	if !caseSensitive {
 		cfg.TableFilter = filter.CaseInsensitive(cfg.TableFilter)
 	}
+	if flags.Lookup(flagWithSysTable) != nil {
```
Comment (3pointer) [inline br/pkg/task/common.go:519]:
```suggestion
```
maybe remove the if?

File: br/pkg/task/common.go [hunk 3]
```diff
@@ -209,6 +210,8 @@ type Config struct {
 
 	CipherInfo backuppb.CipherInfo `json:"-" toml:"-"`
 
+	// determines whether enable restore sys table on default, see fullClusterRestore in restore/client.go
+	WithSysTable bool `json:"with-sys-table" toml:"with-sys-table"`
```
Comment (3pointer) [inline br/pkg/task/common.go:214]: I think only restore need this config, so you can move it to `RestoreCommonConfig`

File: br/pkg/task/common.go (hunk 2 above)
Comment (D3Hunter) [reply br/pkg/task/common.go:519]: restore db/table don't has this flag

**Comment by D3Hunter** [comment]
ci pass
https://ci.pingcap.net/blue/organizations/jenkins/br_ghpr_unit_and_integration_test/detail/br_ghpr_unit_and_integration_test/12662/pipeline

File: br/pkg/task/common.go [hunk 4]
```diff
@@ -310,6 +311,14 @@ func DefineFilterFlags(command *cobra.Command, defaultFilter []string, setHidden
 	}
 }
 
+func DefineWithSysFlags(command *cobra.Command, setHidden bool) {
```
Comment (3pointer) [inline br/pkg/task/common.go:314]: this function can be move into `DefineRestoreCommonFlags`

**Review by 3pointer — COMMENTED** [review]
rest LGTM

**Review by WangLe1321 — APPROVED** [review]

File: br/pkg/task/common.go (hunk 4 above)
Comment (D3Hunter) [reply br/pkg/task/common.go]: [fix comments](https://github.com/pingcap/tidb/pull/36652/commits/1e976ab64458e936afdff84d375a5a7c1b049a53)

Comment (3pointer) [reply br/pkg/task/common.go]: it seems `--with-sys-table` should be a hidden flag

**Review by 3pointer — APPROVED** [review]

**Review by lance6716 — APPROVED** [review]

**Comment by D3Hunter** [comment]
ci pass
https://ci.pingcap.net/blue/organizations/jenkins/br_ghpr_unit_and_integration_test/detail/br_ghpr_unit_and_integration_test/12673/pipeline

**Pull request merged**